syft/pkg/cataloger/golang/licenses.go

```diff
@@ -138,3 +163,79 @@ func processCaps(s string) string {
 		return "!" + strings.ToLower(s)
 	})
 }
+
+func getModule(progress *event.GenericProgress, proxies []string, moduleName, moduleVersion string) (fsys fs.FS, err error) {
+	for _, proxy := range proxies {
+		u, _ := url.Parse(proxy)
+		if proxy == "direct" {
+			fsys, err = getModuleRepository(progress, moduleName, moduleVersion)
+			continue
+		}
+		switch u.Scheme {
+		case "https", "http":
+			fsys, err = getModuleProxy(progress, proxy, moduleName, moduleVersion)
+		case "file":
+			p := filepath.Join(u.Path, moduleName, "@v", moduleVersion)
+			progress.SetValue(fmt.Sprintf("file: %s", p))
+			fsys = os.DirFS(p)
+		}
+		if fsys != nil {
+			break
+		}
+	}
+	return
+}
+
+//nolint:gosec
+func getModuleProxy(progress *event.GenericProgress, proxy string, moduleName string, moduleVersion string) (fs.FS, error) {
+	u := fmt.Sprintf("%s/%s/@v/%s.zip", proxy, moduleName, moduleVersion)
+	progress.SetValue(u)
+	// get the module zip
+	resp, err := http.Get(u)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusOK {
+		u = fmt.Sprintf("%s/%s/@v/%s.zip", proxy, strings.ToLower(moduleName), moduleVersion)
+		progress.SetValue(u)
+		// try lowercasing it; some packages have mixed casing that really messes up the proxy
+		resp, err = http.Get(u)
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = resp.Body.Close() }()
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("failed to get module zip: %s", resp.Status)
+		}
+	}
+	// read the zip
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return zip.NewReader(bytes.NewReader(b), resp.ContentLength)
+}
+
+func getModuleRepository(progress *event.GenericProgress, moduleName string, moduleVersion string) (fs.FS, error) {
+	repoName := moduleName
+	parts := strings.Split(moduleName, "/")
+	if len(parts) > 2 {
+		repoName = fmt.Sprintf("%s/%s/%s", parts[0], parts[1], parts[2])
+	}
+	progress.SetValue(fmt.Sprintf("git: %s", repoName))
+	f := memfs.New()
+	buf := &bytes.Buffer{}
+	_, err := git.Clone(memory.NewStorage(), f, &git.CloneOptions{
```

Comment:
a fs.FS should be passed by the caller instead of `memory.NewStorage()` being used directly. (I've found these kinds of considerations are usually cheap to do during initial development but potentially harder to bolt on later)

Reply:
You mean a temp dir? The idea was for this to stay in memory until it's written out

Reply:
sorry, retracting this, I'm looking at the wrong interface